Add SpectralCoord class to astropy.coordinates

astrofrog opened the pull request:
### About

This pull request implements a new class, ``SpectralCoord`` which aims to represent spectral coordinates (wavelengths, frequencies, energies, velocities) and also encapsulates information about observer and target. This starts from the implementation @nmearl wrote in specutils. Also implemented here is a ``SpectralQuantity`` class with default spectral equivalencies as well as a mechanism for keeping track of the doppler convention and rest value.

The motivation behind this class was:

* to provide a high-level object for use in APE 14 spectral WCSes when calling ``WCS.pixel_to_world`` - currently we returned a Quantity with no meta-data, whereas now we can let the user know what frame of reference the spectral coordinate is for, the direction of the target, and Doppler rest frequencies/wavelengths and conventions used for spectral axes in velocity (this is all information that is part of the FITS WCS standard)

* to provide a class that could be used in specutils to represent spectral axes (in practice specutils will subclass SpectralCoord into SpectralAxis to provide axis-specific functionality too)

Developers are welcome to start reviewing this PR, but note that there are a few remaining to-dos listed lower down, so this is not at the 'final draft' stage. If you have limited time, you could consider just reviewing the docs to see if you are happy with the API.

### Changes to astropy.coordinates

The first set of changes here is to astropy.coordinates - we add a new class ``SpectralCoord`` which is a quantity sub-class with metadata and transformation methods. This also includes a new narrative docs page to describe it. This documentation page is not complete in the sense that some of the functionality is not described, but this provides a starting point, and what's there currently is ready for review.

Note that ``SpectralCoord`` is deliberately a Quantity sub-class rather than using a more complex infrastructure like ``SkyCoord`` with many layers of classes. Since we only care about a single coordinate here, this was found to be the simplest API, and also provides backward-compatibility with the current use of the APE 14 FITS WCS (which previously returned Quantity). We do use coordinate frames/SkyCoord to represent observers and targets though.

I have included a test that does detailed numerical checks against SLALIB - I plan to include more details about how this was done. For a reason I don't understand our galactocentric results are different, and I will be delving into that. For other frames, we typically get ~20m/s agreement for now.

### Changes to astropy.units

I've added a ``SpectralQuantity`` class which implements a ``Quantity`` subclass that automatically includes spectral equivalencies, and keeps track of the doppler convention and rest value.

### Changes to astropy.wcs.wcsapi

The second set of changes is to the APE 14 FITS WCS implementation - here we now return ``SpectralCoord`` for ``pixel_to_world`` - note that we stilll accept ``Quantity`` as input for ``world_to_pixel`` but in this case we give a warning that there is no observer defined, so no velocity transformation will be carried out.

One big caveat is that for now, the target on the ``SpectralCoord`` is set to the reference celestial coordinate in the case of spectral cubes. For cubes with large fields of view, this won't be quite right away from the reference coordinate. But this is already better than the current situation of returning Quantity. So I am planning to add a caveat about this to the APE 14 WCS docs. Properly implementing this may be tricky because it requires interpreting the SSYSOBS keyword which many cubes lack, and in principle this means that even for an LSRK cube the spectral coordinate is actually intended to be constant across slices in the topocentric frame, which means the spectral axis is correlated with the celestial axes. I think this is a little known fact and many cubes may be produced not knowing this assumption, which I think is why the current implementation may actually reflect the way most people use cubes (and pragmatically, it's also a much easier starting point).

### git history

The git history was extracted from specutils (where an initial implementation of this class was added), and I've taken care to preserve the PR merge history so that we have references to the original PRs. I can also remove these merge commits if people would prefer, but I think it's quite nice to be able to see the original chunks of work that were done. I'm happy to go with majority opinion!

### Remaining to-dos

* [x] Make sure the sign of the transformations is correct
* [x] Include code to generate reference rv file, and provide details of where this comes from
* [x] Include ASDF schema and tag
* [x] Fix tests
* [x] Add changelog entry
* [x] Add test for ASDF serialization
* [x] Try and make some improvements to API based on user feedback
* [x] Write documentation for astropy.wcs changes
* [x] Fix xfailed ASDF serialization tests
* [x] Write what's new entry
* [x] Clean up implementation

Things that will be done later after this PR is merged
* [ ] <s>Add docs for ``SpectralQuantity`` </s> we discussed below that for now we can avoid overly advertising this, and leaving it to future to decide if we want to.
* [ ] <s> Add higher accuracy tests (currently limited by the starlink output precision)</s> https://github.com/astropy/astropy/issues/10248

bsipocz commented:
history looks OK to me, except that I don't think we need the merge commit history as those are empty anyway and the content in the commits before the merges anyway with proper author attributions.

pep8speaks commented:
Hello @astrofrog :wave:! It looks like you've made some changes in your pull request, so I've checked the code again for style.








* In the file [`astropy/coordinates/tests/test_spectral_coordinate.py`](https://github.com/astropy/astropy/blob/a1ec9b4c00e12ef7beaf43a1ba32603a951f8584/astropy/coordinates/tests/test_spectral_coordinate.py):

> [Line 61:1](https://github.com/astropy/astropy/blob/a1ec9b4c00e12ef7beaf43a1ba32603a951f8584/astropy/coordinates/tests/test_spectral_coordinate.py#L61): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 148:1](https://github.com/astropy/astropy/blob/a1ec9b4c00e12ef7beaf43a1ba32603a951f8584/astropy/coordinates/tests/test_spectral_coordinate.py#L148): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 376:1](https://github.com/astropy/astropy/blob/a1ec9b4c00e12ef7beaf43a1ba32603a951f8584/astropy/coordinates/tests/test_spectral_coordinate.py#L376): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 0

bsipocz commented:
> The git history was extracted from specutils (where an initial implementation of this class was added), and I've taken care to preserve the PR merge history so that we have references to the original PRs. I can also remove these merge commits if people would prefer, but I think it's quite nice to be able to see the original chunks of work that were done. I'm happy to go with majority opinion!

I feel that the commits should be enough, but checking our current master, I see that we had the PR merges for the wcsaxes migration, too, so I can come around for either way.

StuartLittlefair commented:
Hi @astrofrog - I'm really sorry I don't have time to comment on this PR beyond reading the documentation. It looks like a massive amount of work!

Based on reading the docs I have some feedback, if it's useful:

1. At heart, `SpectralCoord` is a `Quantity`. Should it really be in coordinates? It confuses me, given that it's not a `Frame` or subclass of `SkyCoord`.

1. I find the API very confusing. It bends my head trying to understand what it means when the target has a `radial velocity` and the `SpectralCoord` also has a radial velocity. But it gets worse, since both the `target` and the `observer` possess frames which will have `obstime`s and other frame attributes. Where am I supposed to set what? If I then convert to units of km/s I'll have a `SpectralCoord` in  km/s and two radial velocities? I just can't sort it out.

1. If the accuracy is only 20 m/s this needs a big warning. It's not accurate enough for a lot of stellar work... You could use the corrections from ```SkyCoord.radial_velocity_correction``` for relevant transformations, but then some would be accurate and others not.

I'm not sure what to suggest about the API, but it seems to me that it's a result of trying to hold the same sort of information at several levels. I wonder if it's worth not trying to get this in for 4.1 and having a much longer think about the structure of the code?

astrofrog commented:
@StuartLittlefair - thanks for taking a look at the docs! In the mean time I was working on revamping the docs so if you could take another look that would be great, but I agree there are some more serious issues, see below.

First I should say that we should only merge this once there is consensus that the approach here is reasonable, and if that means we miss v4.1, then that's fine of course, but if we can converge before the end of next week (the updated deadline for *merging* PRs) then that would be great.

Regarding your comments:

> At heart, SpectralCoord is a Quantity. Should it really be in coordinates? It confuses me, given that it's not a Frame or subclass of SkyCoord.

I personally would be equally happy to call it ``SpectralQuantity`` - the rationale for it going in astropy.coordinates is that it can be used to represent absolute spectral coordinates and is heavily tied to the astropy.coordinates machinery. However I don't have strong feelings if people did have another suggestion.

> I find the API very confusing. It bends my head trying to understand what it means when the target has a radial velocity and the SpectralCoord also has a radial velocity. But it gets worse, since both the target and the observer possess frames which will have obstimes and other frame attributes. Where am I supposed to set what? If I then convert to units of km/s I'll have a SpectralCoord in km/s and two radial velocities? I just can't sort it out.

I showed this to another user and got the exact same reaction, so I think we do indeed have an issue here at the moment. Fundamentally, when the quantity itself is a velocity, it really is just a proxy for an absolute frequency relative to a rest frequency. But having that and ``radial_velocity`` be shown at the same time is confusing. For now I've updated the docs to try and make that as clear as possible, but I think for example we would need to consider renaming ``radial_velocity`` to something like ``systemic_velocity`` or ``observer_to_target_velocity`` as ``radial_velocity`` is too ambiguous.

> If the accuracy is only 20 m/s this needs a big warning. It's not accurate enough for a lot of stellar work... You could use the corrections from SkyCoord.radial_velocity_correction for relevant transformations, but then some would be accurate and others not.

The precision at the moment is actually limited by the output values from Starlink which are truncated! Do you have ideas for generating more exact reference values besides ``radial_velocity_correction``?

In general I wonder whether it would make sense to actually move out some of the computation from ``SpectralCoord`` and instead focus on making it so that ``radial_velocity_correction`` can have support for other common frames such as the ones implemented here, or any other arbitrary frames?

> In the above we call the method in_observer_velocity_frame, but the supplied argument isn’t a frame at all, but an object moving in a different frame to the frame we want. Finally, the returned object still has an observer in the ITRS frame? It’s really hard to work out what I just did, even reading the docs.

Yes this is especially confusing in the ``__repr__`` - essentially what we are doing here is keeping the observer at the same position but updating the velocities so that the observer is co-moving with the target.

I'll have a think about the API you propose, thanks! Fundamentally, one requirement of the API is that WCS.pixel_to_world can return an object that is aware of both observer location, target or direction vector towards a target, and velocity frame of reference.

StuartLittlefair commented:
@astrofrog I could see changes in the docs as I was editing my comment. I’ll take another look tomorrow UK time if that’s OK, but I’m signing off for now; I’ve had a long day!

I think if a requirement for WCS.pixel_to_world is to pass one object which is aware of all those things then something like the current API is essential. I actually have a suggestion which might go a long way towards making the API clearer without major changes.

Why not remove radial_velocity and redshift from the repr, and just let users access them as properties. Also, hide properties that are `None` from the repr. Finally, clearly document what the radial velocity property means when more than one of the observer, target and reference frame have a velocity.

astrofrog commented:
> Why not remove radial_velocity and redshift from the repr, and just let users access them as properties. Also, hide properties that are None from the repr. Finally, clearly document what the radial velocity property means when more than one of the observer, target and reference frame have a velocity.

I think this is a very good suggestion - I do think the ``__repr__`` is the most confusing aspect in the current API. And yes if you could review again tomorrow that would be great - thanks! :smile:

mhvk commented:
@astrofrog - I also only had a quick look and very, very much like the idea!

But I also am wondering whether one is not in fact stacking too much into a single class. I think there is actually a good analogy with `SkyCoord` in that you have:
1. Essentially a `Quantity` (analogous to something in between `Angle` and `Representation`) with easy conversion between units.
2. Some reference frame (like `BaseFrame`)
3. And a superstructure that binds them together (`SkyCoord`).

My suggestion would be to separate out at least the part where equivalencies are now baked in - in analogy with functional units, it may be possible to define a generic `SpectralUnit` which has these equivalencies built in. But really I should have a more detailed look; I do appreciate your wish to remain backwards compatible with the `Quantity` that was produced previously...  I'll try next week - right now, I have to get a PR of my own in...

astrofrog commented:
@mhvk - one possible option in terms of code organization is to make a ``SpectralQuantity`` which is the part with the equivalencies built in, and that could live in ``astropy.units``, and then have ``SpectralCoord`` be a sub-class of ``SpectralQuantity``. This would at least avoid having too much code in a single class. We could go the route of having ``SpectralCoord`` contain a ``SpectralQuantity``, but unlike for ``SkyCoord`` where might might be storing up to 6 quantities (positions and velocities) here we should only have to store one, so nesting the classes may be overkill and may make the API less simple to users.

astrofrog commented:
@StuartLittlefair - just FYI I'll be pushing a few more changes that hopefully improve the API and ``__repr__`` today, so you may want to hold off on re-reviewing until I've done that.

astrofrog commented:
@StuartLittlefair - I've pushed some changes, so it would be good if you could now take a fresh look!

StuartLittlefair commented:
Hi @astrofrog 

I'm still only looking at the API and haven't looked at the code at all yet. First comments is that the new repr makes it *much* easier to understand what's going on, so thanks for that. 

The bits I still found challenging to understand all centre on the `with_observer_in_velocity_frame_of` method. Actually, it became much easier to understand once I understood that we are actually changing the `observer`s velocity **as well as** the observer's reference frame. I think if we could make these operations simpler to understand the API would be pretty intuitive.

Might it help to decouple setting the reference frame of the observer, from changing the observer's velocity? I can think of lots of ways you can do so, but one way would be to replace

```python
new = sc_ttau.with_observer_in_velocity_frame_of(sc_ttau.target)
```

with

```python
new = sc_ttau.set_observer_velocity(sc_ttau.target, match_reference_frame=True)
```

or you could have different methods:

```python
new = sc_ttau.set_observer_velocity(sc_ttau.target).set_observer_frame(sc_ttau.target.frame)
```

astrofrog commented:
> It became much easier to understand once I understood that we are actually changing the observers velocity as well as the observer's reference frame

Before my latest changes I was actually only changing the velocities but keeping the frame the same, which I thought was a bit confusing in terms of the ``__repr__``, hence the new behavior - but we could indeed consider splitting that into two API calls. Currently you can change the velocities without changing the actual frame used by specifying ``preserve_observer_frame=True`` in ``with_observer_in_velocity_frame``. Things would be easier if the coordinates machinery allowed the velocities frame to be different from the location frame to some extent.

astrofrog commented:
Just a quick note that I am currently splitting out ``SpectralCoord`` into ``SpectralCoord`` and ``SpectralQuantity`` following @StuartLittlefair and @mhvk's suggestion!

astrofrog commented:
I've now added a ``SpectralQuantity`` class to ``astropy.units`` which ``SpectralCoord`` then inherits from. The former deals with any spectral conversions and the Doppler convention/rest bookkeeping.

mhvk commented:
@astrofrog - I hope to review in the next few days. The split certainly sounds good, though I'm not sure anything needs to be in `units` (after all, angles are also in `coordinates`); documentation at least may be more logically together in one place. Anyway, will have a look!

astrofrog commented:
@mhvk - I'm happy to move things around as needed - for now I've put it in separate files so that at least moving things is easy if we agree on the rest :)

taldcroft commented:
@astrofrog @mhvk - it's not necessary or realistic for 4.1, but I'm assuming it would be possible and desirable to allow this class to be used as a Table mixin column and be serialized to FITS, ECSV, HDF5?

astrofrog commented:
@taldcroft - indeed this would be very nice. Might the mix-in already work since it's a Quantity sub-class?

taldcroft commented:
@astrofrog - so this would need an `Info` class that provides details of serialization.  It might be easy or it might be not so easy.  E.g. in https://github.com/astropy/astropy/pull/10210#issuecomment-620527211 it starts get a little tricky to fully serialize `Representation` objects because of differentials.  So the question is about the layers of attributes that comprise your new class and which of those can be array-valued.  Flat is easier, hierarchical is harder.

astrofrog commented:
LSRK and LSRD are now proper frames, defined alongside LSR, which allows the 'lsrk' and 'lsrd' string aliases to be used. The galactocentric, local group, and CMB frames are for now just defined for FITS-WCS - if/once this PR gets merged, I'll have a think about the best way to code up those frames in astropy.coordinates (and whether we even want to), but I think those are less critical initially.

astrofrog commented:
This is now ready for detailed reviews. Following discussions with @adamginsburg and @StuartLittlefair, I changed the name of the method to change the observer velocities to ``with_observer_velocities``.

Please feel free to review docs, implementation, and take this for a spin if you can to see if you are happy with the API!

keflavich commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,850 @@
+__doctest_skip__ = ['SpectralCoord.*']
+
+
+def _velocity_to_refshift(velocity):
```
redshift?

astrofrog replied:
Yes, fixed!

keflavich commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,850 @@
+    """
+    Coordinate object representing spectral values.
+
+    The `SpectralCoord` class is new in Astropy v4.1 and should be considered
```
should this be a `..note::` or similar rst formatted block?

astrofrog replied:
Yes, fixed!

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        """
+        return SpectralCoord, True
+
+    @staticmethod
+    def _target_from_observer(observer, radial_velocity):
```
```suggestion
    @staticmethod
    @u.quantity_input(radial_velocity=u.km/u.s)
    def _target_from_observer(observer, radial_velocity):
```

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        tot_rv = radial_velocity  # + observer_icrs.radial_velocity
+
+        target = (observer_icrs.cartesian.without_differentials() + drep).with_differentials(
+            CartesianDifferential([obs_vel.d_x + tot_rv,
+                                   obs_vel.d_y.to(tot_rv.unit),
+                                   obs_vel.d_z.to(tot_rv.unit)]))
```
```suggestion
            CartesianDifferential([obs_vel.d_x + radial_velocity,
                                   obs_vel.d_y.to(radial_velocity.unit),
                                   obs_vel.d_z.to(radial_velocity.unit)]))
```

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        return self.__class__(**default_kwargs)
+
+    @property
+    def quantity(self):
```
I'd prefer a method, e.g., `as_quantity()`, or nothing (i.e. use `view()` directly) but don't feel strongly.

astrofrog replied:
I prefer ``as_quantity`` too - changed!

astrofrog replied:
Changed back following @mhvk's review :grin:

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+
+    def _project_velocity_and_shift(self, init_vel, fin_vel, line_of_sight_unit_vec):
+        """
+        Calculated the velocity projection given two vectors.
```
```suggestion
        Calculate the velocity projection given two vectors.
```

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+
+        return self._apply_relativistic_doppler_shift(delta_vel)
+
+    def with_observer_velocity(self, frame, preserve_observer_frame=False):
```
Hm, this name is confusing to me, but I don't immediately have a better suggestion...

astrofrog replied:
Ha, this has been the subject of many discussions so far and I've not yet found a name that satisfies everyone! Other names it has had though the commit history are: ``in_observer_velocity_frame``, ``with_observer_in_velocity_frame_of``, and I've also considered ``with_observer_stationary_relative_to``. Naming things is hard :stuck_out_tongue:

mhvk replied:
My suggestion (see earlier) is `transform_observer_to()`, though now that I look at it more closely, perhaps it doesn't make so much sense. The more general `transform_to` might still work, though, and is nicely analogous to `SkyCoord`.

StuartLittlefair replied:
The reason I originally suggested something like the current name is that this routine is leaving the position of the observer unchanged, but changing the observers velocity *and* the frame in which it is represented. 

Suggestions along the `transform*` route suggest the observer remains unchanged which is misleading. 

I also suggested not changing the observer’s frame here but I think the objection is that it produces output that is correct but surprising (v large ITRS velocities)

mhvk replied:
Yikes, I'm seeing how this is indeed tricky.

StuartLittlefair replied:
Of the many suggestions so far I think `with_observer_stationary_relative_to` is verbose but clearest. Maybe just `stationary_relative_to` would be OK?

I have a suggestion - not sure how big a change it would be. The idea is to make it clearer what is happening by changing the arguments. So instead of passing in a frame instance whose position argument is (perhaps confusingly) ignored, pass the velocity and frame *type* separately.

```python
sc_tttau_bary = sc_ttau.stationary_relative_to('icrs', velocity=vel)
sc_tttau_bary = sc_ttau.stationary_relative_to(coord.ICRS, vel)
```

The velocity argument would be anything that can be converted to a ```CartesianDifferential``` and would be *optional* so:

```python
sc_tttau_bary = sc_ttau.stationary_relative_to('icrs')
```

would give a ```SpectralCoord``` with an observer stationary w.r.t to the barycenter. 
This looks cleaner and more intuitive to my eyes, but the downside would be that getting the ```SpectralCoord``` relative to the rest frame of the target would be a lot clunkier:

```python
sc_ttau_rest = sc_ttau.stationary_relative_to(
    sc_ttau.target.frame, 
    sc_ttau.target.cartesian.differentials['s']
)
```

astrofrog replied:
Thanks for the new idea! See my latest comment on the PR.

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+
+        return new_coord
+
+    def with_los_shift(self, target_shift=None, observer_shift=None):
```
Elsewhere, `line of sight` is always `radial`. Although very verbose, perhaps this should be 
```suggestion
    def with_radial_velocity_shift(self, target_shift=None, observer_shift=None):
```

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        """
+        Apply a velocity shift to this spectral coordinate. The shift
+        can be provided as a redshift (float value) or radial velocity
+        (quantity with physical type of 'speed').
```
```suggestion
        (`~astropy.units.Quantity` with physical type of 'speed').
```

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+            raise ValueError("Both an observer and target must be defined "
+                             "before applying a velocity shift.")
+
+        for arg in [x for x in [target_shift, observer_shift] if x is not None]:
+            if isinstance(arg, u.Quantity) and \
+                    arg.unit.physical_type not in ['speed', 'dimensionless']:
+                raise u.UnitsError("Argument must have unit physical type "
+                                   "'speed' for radial velocty or "
+                                   "'dimesionless' for redshift.")
+
+        # The target or observer value is defined but is not a quantity object,
+        #  assume it's a redshift float value and convert to velocity
+        if isinstance(target_shift, (float, int)) or \
+                isinstance(target_shift, u.Quantity) and \
+                target_shift.unit.physical_type == 'dimensionless':
+            target_shift = _redshift_to_velocity(target_shift)
+
+        if isinstance(observer_shift, (float, int)) or \
+                isinstance(observer_shift, u.Quantity) and \
+                observer_shift.unit.physical_type == 'dimensionless':
+            observer_shift = _redshift_to_velocity(observer_shift)
```
Some of this could probably be cleaned up with the `quantity_input()` decorator added above.

adrn replied:
(but even if not, you could clean this up by first passing values in to `u.Quantity()` - floats will become dimensionless quantities)

astrofrog replied:
Done

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        new_coord = self._change_observer_to(observer)
+
+        return new_coord
+
+    def with_los_shift(self, target_shift=None, observer_shift=None):
```
```suggestion

    @u.quantity_input(target_shift=['speed', u.dimensionless_unscaled],
                  observer_shift=['speed', u.dimensionless_unscaled])
    def with_los_shift(self, target_shift=None, observer_shift=None):
```
You may not want this, because with this a float value would not be allowed...

astrofrog replied:
As above, I think indeed we don't want that for now

adrn commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        new_coord = self._change_observer_to(observer)
+
+        return new_coord
+
```
```suggestion

    @u.quantity_input(target_shift=[u.km/u.s, u.dimensionless_unscaled],
                      observer_shift=[u.km/u.s, u.dimensionless_unscaled])
```
You may not want this, because with this a float value would not be allowed...

astrofrog replied:
I agree that we don't want that for now since we do want to allow plain Python scalars *but* we could in future patch ``quantity_input`` to allow ints/floats to be passed if u.dimensionless_unscaled is acceptable?

adrn replied:
I got inspired :) https://github.com/astropy/astropy/pull/10232

mhvk commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+    def to_value(self, *args, **kwargs):
+        return self.to(*args, **kwargs).value
+
+    def _apply_relativistic_doppler_shift(self, velocity):
```
I don't think one should define private methods for use of classes in a different scope; I think it may make more sense to move it up to where it is used, or make it a separate function.

mhvk commented on astropy/units/tests/test_spectral_quantity.py:
```diff
@@ -0,0 +1,166 @@
+        sq13 = SpectralQuantity(200 * u.km / u.s, doppler_convention='optical', doppler_rest=1 * u.GHz)
+        sq14 = sq13._apply_relativistic_doppler_shift(300 * u.km / u.s)
+        assert_quantity_allclose(sq14, 500.350827 * u.km / u.s)
+        assert sq14.doppler_convention == 'optical'
```
(Given previous comment, this maybe shouldn't be here)

Also test adding `velocity=0.999*c` twice.

astrofrog replied:
Fixed

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+__doctest_skip__ = ['SpectralCoord.*']
+
+
+def _velocity_to_redshift(velocity):
```
I agree but maybe not in this PR - I think the best place for this would be an equivalency?

astrofrog replied:
I opened an issue in https://github.com/astropy/astropy/issues/10240

astrofrog commented on astropy/coordinates/builtin_frames/lsr.py:
```diff
@@ -158,3 +159,100 @@ def galacticlsr_to_galactic(lsr_coord, galactic_frame):
+
+
+# The following velocities are defined in FK4 B1900
+V_BARY_GORDON1975 = r.CartesianDifferential([0, -10 * 3 ** 0.5, 10.]*u.km/u.s)
```
I went with the simplifications below

astrofrog commented on astropy/coordinates/builtin_frames/lsr.py:
```diff
@@ -158,3 +159,100 @@ def galacticlsr_to_galactic(lsr_coord, galactic_frame):
+
+@frame_transform_graph.transform(AffineTransform, ICRS, LSRK)
+def icrs_to_lsrk(icrs_coord, lsr_frame):
+    v_bary_fk4 = FK4(V_BARY_GORDON1975.to_cartesian(), equinox='B1900')
```
Done!

astrofrog commented on astropy/coordinates/builtin_frames/lsr.py:
```diff
@@ -158,3 +159,100 @@ def galacticlsr_to_galactic(lsr_coord, galactic_frame):
+
+@frame_transform_graph.transform(AffineTransform, ICRS, LSRD)
+def icrs_to_lsrd(icrs_coord, lsr_frame):
+    v_bary_gal = Galactic(V_BARY_DELHAYE1965.to_cartesian())
```
Done

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        # The quantity machinery will drop the unit because type(value) !=
+        # SpectralCoord when passing in a Quantity object. Reassign the unit
+        # here to avoid this.
+        if isinstance(value, u.Quantity) and unit is None:
```
It turns out this wasn't needed (just added a test)

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+        # for internal purposes.
+        if observer is None:
+            if target is None:
+                observer = ICRS(ra=0 * u.degree, dec=0 * u.degree,
+                                pm_ra_cosdec=0 * u.mas/u.yr, pm_dec=0 * u.mas/u.yr,
+                                distance=0 * u.pc, radial_velocity=0 * u.km/u.s)
```
Maybe it could be a helper function just for the test suite?

mhvk replied:
As part of having classes get `__array_function__`, one can let `np.zeros_like(...)` become possible (which, these days, includes being able to override the shape).

astrofrog replied:
I opened an issue in https://github.com/astropy/astropy/issues/10241

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,851 @@
+                                pm_ra_cosdec=0 * u.mas/u.yr, pm_dec=0 * u.mas/u.yr,
+                                distance=0 * u.pc, radial_velocity=0 * u.km/u.s)
+            else:
+                if radial_velocity is None:
+                    radial_velocity = 0 * u.km/u.s
```
Done

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+
+        obj = super().__new__(cls, value, unit=unit, **kwargs)
+
+        # The quantity machinery will drop the unit because type(value) !=
```
Removed

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+
+        # If we're initializing from an existing SpectralQuantity, keep any
+        # parameters that aren't being overridden
+        if isinstance(value, SpectralQuantity):
```
Done

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+        if self._doppler_rest is None:
+            self._doppler_rest = value
+        else:
+            raise ValueError("doppler_rest has already been set, and cannot "
```
Fixed

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+        if self._doppler_convention is None:
+            self._doppler_convention = value
+        else:
+            raise ValueError("doppler_convention has already been set, and cannot "
```
Done

astrofrog commented:
@adrn @mhvk - thanks for your reviews! I've started to address your comments, but will finish tomorrow morning. In the mean time let me know if you have any other comments!

astrofrog commented:
@mhvk - just a quick note on:

> My tendency would also be to keep this in coordinates for now, as I think it is most similar to Representation and Angle.

I would have thought that ``SpectralQuantity`` was not very tied to coordinates, since it really just encapsulates a few spectral units and equivalencies whereas Representations are still for some form of 3-d coordinates, and Angle has a lot of parsing things related to coordinates. However, I'm happy to move it to astropy.coordinates if you do think it makes more sense there. We could even not expose it publicly initially. Do any of the other coordinates/units maintainers have any preference? (note that I'm specifically asking about ``SpectralQuantity``, not ``SpectralCoord``.

mhvk commented:
I think not exposing it overly for now, i.e., treating it as a bit of an implementation detail for `SpectralCoord`, might be a good idea.  But really I cannot say too clearly until I have looked in more detail at the rest...

astrofrog commented:
@mhvk - ok, I'll keep it where it is for now until you've had a chance to review the rest :)

mhvk commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+The ``radial_velocity``, which is the velocity offset between observer and
+target, is now zero.
+
+|SpectralCoord| is very versatile and can be used to represent any spectral
```
Maybe: "is intended to be versatile and be useful for representing ..." (I always feel it should be up to the reader to judge qualities)

astrofrog replied:
Fixed

mhvk commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+|SpectralCoord| is very versatile and can be used to represent any spectral
+values - not just the x-axis of a spectrum, but also for example the
+frequencies of spectral features. For example, if we now consider that we found a
+spectral feature that appears to have three features at the following frequencies
```
double "feature" - maybe "spectral feature that has components at the following three ..."

astrofrog replied:
Done

mhvk commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+======================
+
+Any valid astropy coordinate frame can be passed to the
+:meth:`~astropy.coordinates.SpectralCoord.with_observer_velocity`
```
So, this suggests one cannot actually just pass in a velocity (still haven't looked at actual code)...

astrofrog replied:
Correct - which was my objection to that name :) I agree the current name is still not right

mhvk commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+                           V=12 km/s, and W=7 km/s in Galactic coordinates
+                           (equivalent to 16.552945 km/s towards l=53.13
+                           and b=25.02) [2]_.
+``'lsr'``                  A more recent definition of the Local Standard
```
If this is more recent, shouldn't it be used in the examples above?

astrofrog replied:
Yes and no - it's more recent but also not necessarily 'standard' yet - LSRD and LSRK have been widely used for years. I actually am not sure it was a good idea to have a plain LSR paper and just base it off a recent paper, because it's by no means a standard, just one study.

mhvk replied:
OK, makes sense. Maybe an idea to even discuss this when you introduce `LSRD`.

astrofrog replied:
I reworded this

mhvk commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+.. [3] Schönrich, R., Binney, J., & Dehnen, W. 2010, MNRAS, 403, 1829
+       `[ADS] <https://ui.adsabs.harvard.edu/abs/2010MNRAS.403.1829S>`__.
+
+.. The following frames are defined in FITS WCS and may be added here in future:
```
How about uncommenting this and giving an example of how one would transform to one of these oneself?

astrofrog replied:
Done - I added an example of the local group velocity frame

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        pos_hat : `BaseRepresentation`
+            Position representation.
+        """
+        d_pos = (target.data.without_differentials() -
```
The subtraction will go through `cartesian` anyway, so perhaps `target.cartesian.without_differentials() - observer.cartesian.without_differentials()`

mhvk replied:
Actually, the rest could be `return d_pos.represent_as(UnitSphericalRepresentation).to_cartesian()`. But slower.

astrofrog replied:
Ok I fixed the subtraction

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        dp_norm = d_pos.norm()
+
+        # Reset any that are 0 to 1 to avoid nans from 0/0
+        dp_norm.ravel()[dp_norm.ravel() == 0] = 1 * dp_norm.unit
```
No need to `ravel`, just `dp_norm[dp_norm == 0] = 1 * dp_norm.unit`

astrofrog replied:
Fixed

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        init_obs_vel = self._calculate_radial_velocity(self.observer, target)
+        fin_obs_vel = self._calculate_radial_velocity(observer, target)
+
+        line_of_sight_unit_vec = self._norm_d_pos(observer, target).xyz
```
I'd really hope it would not be necessary to take the quantity here, i.e., do without `.xyz`. I think with my suggested edit to `_calculate_radial_velocity` it may not be necessary.

astrofrog replied:
Done

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        if as_scalar:
+            return vel_mag
+        else:
+            return vel_mag * pos_hat.xyz
```
I'd just return `vel_mag * pos_hat` to keep using representations. Saves you trouble later!

astrofrog replied:
Done

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+
+        # Project the velocity shift vector onto the line-of-sight vector
+        # between the target and the new observation frame.
+        init_proj_vel = np.dot(init_vel, line_of_sight_unit_vec) * line_of_sight_unit_vec
```
If one keeps this in terms of representations, this is just `line_of_sight_unit_vec.dot(init_vel) * line_of sight_unit_vec` (and no need for `NUMPY_LT_1_17`)

astrofrog replied:
Done

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        # the dot product to determine this.
+
+        diff_vel = fin_vel - init_proj_vel
+        delta_vel = np.dot(diff_vel, line_of_sight_unit_vec)
```
as above, one would have `line_of_sight_unit_vec.dot(diff_vel)`

astrofrog replied:
This currently returns a vector as opposed to a scalar, so it seems I have to then access ``.norm()``?

astrofrog replied:
Done

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        # zero. In this case, set a tolerance for the final velocity; if it's
+        # below this tolerance, assume the delta velocity is essentially nan.
+
+        fin_vel_mag = np.linalg.norm(fin_vel)
```
This would become `fin_vel.norm()`

astrofrog replied:
Done

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+                              0 * u.m / u.s, 0 * u.m / u.s, 0 * u.m / u.s,
+                              representation_type='cartesian',
+                              differential_type='cartesian')
+        elif isinstance(frame, str):
```
Maybe swap branches so that the initialization with `0` only  has to be done once?
```
if hasattr...
    ...
else:
    if instance(frame, str):
        ...
    if isinstance(frame, type):
        ...
```

astrofrog replied:
Done

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+
+    def with_radial_velocity_shift(self, target_shift=None, observer_shift=None):
+        """
+        Apply a velocity shift to this spectral coordinate. The shift
```
I'd have a break after the first sentence, so that the second becomes a note.

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+
+        for arg in [x for x in [target_shift, observer_shift] if x is not None]:
+            if isinstance(arg, u.Quantity) and \
+                    arg.unit.physical_type not in ['speed', 'dimensionless']:
```
Maybe just `args.unit.is_equivalent((u.dimensionless_unscaled, u.km/u.s))`? Allows possible user-defined equivalencies that are in play.

astrofrog replied:
Done!

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        target_icrs = self._target.transform_to(ICRS)
+        observer_icrs = self._observer.transform_to(ICRS)
+
+        target_shift = 0 * u.km / u.s if target_shift is None else target_shift
```
Maybe have this as an `else` clause at the end of `if target_shift is not None` above?

astrofrog replied:
Done

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+
+        new_target = target_icrs.realize_frame(
+            target_icrs.cartesian.with_differentials(
+                CartesianDifferential(target_velocity.xyz))).transform_to(
```
I think `target_velocity` is always a `CartesianDifferential`, no?

astrofrog replied:
No, it seems it is a ``CartesianRepresentation`` because above we add a differential to a representation so we get a representation. So I still need the explicit conversion to a differential.

mhvk commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        init_obs_vel = self._calculate_radial_velocity(observer_icrs, target_icrs)
+        fin_obs_vel = self._calculate_radial_velocity(new_observer, new_target)
+
+        line_of_sight_unit_vec = self._norm_d_pos(observer_icrs, target_icrs).xyz
```
See comments above on keeping things as representations.

astrofrog replied:
I've changed it so that ``_norm_d_pos`` now returns a ``CartesianRepresentation``.

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+           doppler_rest=None,
+           doppler_convention=None):
+        """
+        Overloaded parent ``to`` method to provide parameters for defining
```
Done

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+        ----------
+        doppler_rest : `~astropy.units.Quantity`, optional
+            The rest value used in the velocity space conversions. Providing
+            the value here will set the value stored on the `SpectralQuantity`
```
Done

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+
+        Parameters
+        ----------
+        doppler_rest : `~astropy.units.Quantity`, optional
```
Done

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+
+            vel_equiv1 = DOPPLER_CONVENTIONS[self._doppler_convention](self._doppler_rest)
+
+            freq = super().to(si.Hz, equivalencies=equivalencies + eq.spectral() + vel_equiv1)
```
Good point, fixed!

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+        # FIXME: need to consider case where doppler equivalency is passed in
+        # equivalencies list, or is u.spectral equivalency is already passed
+
+        if doppler_rest is None:
```
Yes the idea is to preserve the preferred Doppler settings even if converting from Hz to AA

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+
+        else:
+
+            additional_equivalencies = eq.spectral()
```
The doppler equivalencies don't hard-code the conversion from some spectral units such as wavenumber, so it's still needed for those cases. We could arguably expand the Doppler equivalencies to include this (in fact, I wonder whether maybe the Doppler equivalencies could be written in a more general way that just uses spectral to convert to a common unit before converting to velocity). Maybe best dealt with in a separate PR though.

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+        Given a `SpectralQuantity` and a velocity, return a new `SpectralQuantity`
+        that is Doppler shifted by this amount.
+
+        Note that the Doppler shift applies is the full relativistic one, so
```
Done

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+                result = squantity
+            else:
+                result = squantity.to(squantity.unit, doppler_convention='relativistic')
+            result = result + velocity
```
Fixed - it's actually a lot simpler to just convert to frequency, apply the shift, and then transform back

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+                result = result.to(squantity.unit, doppler_convention=squantity.doppler_convention)
+            return result
+        else:
+            raise TypeError(f"Unexpected units in velocity shift: {squantity.unit}")
```
Done

astrofrog commented on astropy/units/tests/test_spectral_quantity.py:
```diff
@@ -0,0 +1,166 @@
+
+from ..spectral_quantity import SpectralQuantity
+
+SPECTRAL_UNITS = (u.GHz, u.micron, u.keV, (1 / u.nm).unit, u.km / u.s)
```
Done

astrofrog commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+energies, as well as equivalent Doppler velocities. While the plain |Quantity|
+class can also represent these kinds of physical quantities, and allow
+conversion via dedicated equivalencies (such as :ref:`u.spectral
+<astropy-units-doppler-equivalencies>` or the :ref:`u.doppler_*
```
Good catch! Fixed

astrofrog commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+===============
+
+By default, unit conversions between spectral units will work without having to
+specify the :ref:`u.spectral <astropy-units-doppler-equivalencies>` equivalency::
```
Fixed

astrofrog commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+For the target, the simplest way is to use the |SkyCoord| class::
+
+    >>> from astropy.coordinates import SkyCoord
+    >>> ttau = SkyCoord('04h21m59.43s +19d32m06.4', frame='icrs',
```
:+1:

astrofrog commented on docs/coordinates/spectralcoord.rst:
```diff
@@ -0,0 +1,530 @@
+which is because the observer is now stationary relative to the barycenter so has
+a significant velocity relative to the surface of the Earth.
+
+We can also transform the frequencies to the LSRD frame of reference::
```
Done

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+                                                              CartesianDifferential).differentials
+
+    if original_icrs.data.differentials:
+        data_with_differentials = original_icrs.data.represent_as(CartesianRepresentation,
```
Good catch, and have also shortened the lines

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+                radial_velocity=None, redshift=None,
+                **kwargs):
+
+        # Enforce subok=True even if already specified
```
I've just removed this

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+        self._doppler_rest = getattr(obj, '_doppler_rest', None)
+        self._doppler_convention = getattr(obj, '_doppler_convention', None)
+
+    def __quantity_subclass__(self, unit):
```
Ok so this is an interesting problem - AA/s is actually a valid spectral unit since it's velocity! For now I've removed ``__quantity_subclass__`` but see the ``test_arithmetic`` test for SpectralQuantity - there is a commented out line that fails because of this. This is beyond by units-fu - what would be the correct way to make it so that changes in unit only work if done through a ``to`` conversion but not arithmetic?

mhvk replied:
Ouch...  For now maybe easiest to just rely on what `SpecificTypeQuantity` does and we can try to fix this later.

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        self._observer = getattr(obj, '_observer', None)
+        self._target = getattr(obj, '_target', None)
+
+    def __quantity_subclass__(self, unit):
```
Removed

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        obs_vel = observer_icrs.cartesian.differentials['s']
+
+        target = (observer_icrs.cartesian.without_differentials() + drep).with_differentials(
+            CartesianDifferential([obs_vel.d_x + radial_velocity,
```
You are correct - I've now added a ``_relativistic_velocity_addition`` function and use it here

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        coord : `~astropy.coordinates.BaseCoordinateFrame`
+            The new frame to be used for target or observer.
+        """
+        if not issubclass(coord.__class__, (BaseCoordinateFrame, FrameMeta)):
```
I'm not sure, but everything seems to work fine without, so removing. @nmearl - let me know if there was a good reason for this!

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+        return self.__class__(**default_kwargs)
+
+    @property
+    def as_quantity(self):
```
Good point - I've changed it back :)

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+    @property
+    def observer(self):
+        """
+        The coordinate frame from which the observation was taken.
```
Fixed

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,842 @@
+
+        d_vel = target_icrs.velocity - observer_icrs.velocity
+
+        vel_mag = np.dot(d_vel.d_xyz, pos_hat.xyz)
```
Changed!

astrofrog commented:
@mhvk - thank you for your review! Your comments are always :100: :smile: 

> @astrofrog - now reviewed the SpectralQuantity part. I think the separation of concerns is very good, though I think _apply_relativistic... should be moved off the class (and have some other in-line comments).

I've moved ``_apply_relativistic_doppler_shift`` to being an internal function in spectral_coordinates.py

> My tendency would also be to keep this in coordinates for now, as I think it is most similar to Representation and Angle.

Done. For now I've left it public, but we could also potentially make it private.

> A larger question is to what extent you want to allow calculations with these; I definitely think that sq / q should give a Quantity if the units no longer are spectral (see in-line comment), but what about sq ± sq? Safest might again be to drop to Quantity (e.g., if both sq are in velocity units, shouldn't addition be relativistic?) but should Hz + A be allowed? (probably not.) Maybe similar to what we do with Longitude and Latitude, using _no_angle_subclass.

I think we want multiplication and division by scalars should return SpectralCoord/SpectralQuantity but I guess all the rest should probably drop down to ``Quantity``. I'll try and work on this now.

> Finally, on the implementation, I wonder a bit whether it might be even better to have a SpectralUnit which carries the equivalencies (including doppler). An advantage is that things like sq.unit.is_equivalent(u.m) could be made to be always True (this is what I've done for the logarithmic units). But I think what you have so far does not preclude moving to such a scheme (if it is in fact a good idea, not sure at all!).

I think that could work, although I'd indeed prefer if we punt on this for now, since the current approach is still valid/useful (I think)

> I think not exposing it overly for now, i.e., treating it as a bit of an implementation detail for SpectralCoord, might be a good idea. But really I cannot say too clearly until I have looked in more detail at the rest...

Do you have any further thoughs on this now that you've seen the rest?

> But I wonder about names and the repr, thinking that ideally we make it more consistent with SkyCoord. In that respect,
> Perhaps the __repr__ should have the actual values at the end, just like SkyCoord, and have observer, target and RV/z beforehand, as, effectivelyM SpectralCoord attributes.
    
I'll try and improve these - though at least if we want to tweak it later, it probably doesn't count as breaking API :)

> The method with_observer_velocity should perhaps be called transform_observer_to. Or, more drastically and perhaps better one could imagine doing everything with a .transform_to, as in
> sc.transform_to(radial_velocity=20*u.km/s) sc.transform_to(observer='icrs')
> sc.transform_to(target='T Tau N')  # not necessarily a crazy transformation.

This is quite a significant change and I think the issue is that it perhaps hides the fact we aren't just tansforming the observer to another frame, we are actually editing its velocities. I'd be in favor of keeping a dedicated method for this for now, and then see during the 4.2 dev cycle if we can come up with anything better? (for now the API is marked as experimental so I think deprecating this would be fair game if needed)

Personally I think the clearest names would be:

* ``with_observer_stationary_relative_to(...)``
* ``with_observer_in_velocity_frame_of(...)``

Other names suggested so far:

* ``observer_in_velocity_frame(...)``
* ``with_observer_velocities(...)``

To me, while verbose, ``with_observer_stationary_relative_to`` is the most explicit option and describes exactly what we are doing - we are using whatever is passed as a frame or coordinate in which or relative to which the observer is made to be stationary.

> p.s. Am also not completely sold yet on SpectralCoord being a Quantity subclass. Maybe behaving like a Quantity is good enough (like TimeDelta). But I really should look first!!

I can see arguments both ways - having it not inherit from Quantity would mean breaking WCS API though for spectral cubes, so if we are on the fence, I'd vote to keep it a subclass.

> Only smaller comments so far. I'm a bit confused about the need for the fake positions, but I guess since that is hidden it is an implementation detail.

I personally also think we could simplify this but didn't want to touch it for now, as indeed it's an implementation detail that can be fixed after feature freeze.

I'm going to work on the ``__repr__`` and improving test coverage for ``SpectralCoord``, but if you have any thoughts on the remaining points above, let me know!

astrofrog commented:
@mhvk - could you take a look at https://github.com/astropy/astropy/pull/10185/commits/23fee5b217031e026f14e88c869c01a66e8dc6f7 to see how I'm handling the arithmetic/ufuncs? Does this seem the right way to do it?

mhvk commented:
@astrofrog, just a quick :+1: to having `SpectralQuantity` in coordinates. I think it is fine to keep it public; it may be useful on its own (and we do warn about API changes).

astrofrog commented:
@mhvk - thanks for the comments on the ``__array_ufunc__`` implementation! I've implemented your suggestions, though I'm having trouble getting the test for ``np.std`` to pass.

Does anyone have any further opinions regarding the name for the method that changes the observer velocities? Given all the various feedback, I'm tempted to suggest we go with something like ``with_observer_stationary_relative_to`` for now, which I think should make it clear what's being done. Would anyone have any objections to this?

astrofrog commented on astropy/units/spectral_quantity.py:
```diff
@@ -0,0 +1,291 @@
+        while negative velocities blueshift the spectral quantity.
+        """
+
+        # NOTE: we deliberately don't keep sub-classes of SpectralQuantity intact
```
No longer relevant since it's not a method

mhvk commented:
@astrofrog - hah, you found a bug! #10245 - `longitude.std()` also returns a `Longitude` instead of the desired `Angle`. I think you can avoid it by treating `_wrap_function()` like you did `__array_ufunc__`, but frankly I think it is also fine to let it be for the moment (ie., `xfail` the test, mentioning #10245)

mhvk commented:
On the name, in-line @StuartLittlefair suggested the shorter `stationary_relative_to`. I think I prefer that but throw out two further ideas anyway: `velocity_relative_to` or just `.to` and work with arguments (probably too overloaded).

I also liked @StuartLittlefair's suggestion of having two arguments `frame, velocity` - note that I think it can be a special case that if `frame` is an instance with a velocity differential, those get used. If so, one would still allow the current signature, but the arguments would help better explain what is actually being done.
 (and a possible different set of arguments, which I

mhvk commented:
p.s. With the two-argument version, one might even just go with `relative_to`?

StuartLittlefair commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+        # Validate the observer and target
+        if observer is not None and not isinstance(observer, (SkyCoord, BaseCoordinateFrame)):
+            raise ValueError("observer must be a sky coordinate or coordinate frame.")
+        if target is not None and not isinstance(target, (SkyCoord, BaseCoordinateFrame)):
```
likewise - use `_validate_coordinate`?

astrofrog replied:
Done

astrofrog commented:
@mhvk @StuartLittlefair - the issue with not having ``observer`` in the name is that it doesn't make it clear that this is something that affects the observer and not the target. I propose that for now we go with the more verbose/explicit ``with_observer_stationary_relative_to`` and we could always consider introducing a more concise name in v4.2 - we might get user feedback about other parts of SpectralCoord anyway, and it is marked as experimental, so I think this could be possible? Based on all the discussions I've had, I've found short names to be the most controversial hence why slightly longer/explicit might be better for now. However, I really like the idea of having the optional explicit velocity to allow simple calls like ``('icrs', velocity=[2, 0, 1] * u.km / u.s])`` and so on, so I've implemented that and added a detailed test of the different ways to call this method.

@mhvk - I've tried to improve the ``__repr__`` to match ``SkyCoord`` better *(in particular enclosing the attributes in brackets and having the data come at the end).

I also removed the ``with_redshift`` and ``with_radial_velocity`` methods, since these names did not make it clear if any conversion/transformation was taking place or not (it wasn't converting, just replacing). Instead I've implemented a ``replicate`` method (replacing the previous private ``_copy``) which mirrors what is present on ``SkyCoord`` and ``Time`` and allows any attributes to be changed without doing conversions/transformations.

I've now added a what's new entry, and created an issue to remind us to add higher accuracy tests that what is here currently (https://github.com/astropy/astropy/issues/10248). I think this would be easier to handle in a separate PR and could be done after feature freeze (with issues fixed during bugfix releases if needed).

At this point, I think this is ready for a final set of reviews, and it would be great if we could get this in today before the feature freeze.

So @mhvk @StuartLittlefair @adrn - since you already did detailed reviews of this (thanks!), could you take one last look and approve if ready (or let me know if there are any remaining issues)

@eteq @nmearl - since there have been some changes since the specutils version, this is your chance to take a look (no fundamental changes, just some API cleanup based on all the detailed reviews above, as well as fixes to the implementation for correctness)

@nden @Cadair - are you happy with the WCS-related changes here?

@keflavich - you mentioned you might try this out - did you get a chance to do this?

StuartLittlefair commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+        #  with just the radial velocity set so new implicit observer/target
+        #  will be created. Otherwise, use the available observer/target
+        #  instances to create the pair.
+        # FIXME: radial_velocity below is wrong
```
is it?

astrofrog replied:
It was and I've now fixed it

StuartLittlefair commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+
+        pos_hat = SpectralCoord._norm_d_pos(observer_icrs, target_icrs)
+
+        d_vel = target_icrs.velocity - observer_icrs.velocity
```
I'm getting a bit brain-dead as I review this quickly, but shouldn't this use the relativistic addition formula?

astrofrog replied:
I think so, and I think there are a few other places where this is needed too. Since I'm short on time in terms of doing this before feature freeze, I've opened an issue about it here:

https://github.com/astropy/astropy/issues/10249

I think it will be easier to focus on that separately from all the other implementation details here next week, and since that is just fixing numerical accuracy rather than a new feature, it can be merged after feature freeze or make it in a bugfix release if needed.

mhvk replied:
@astrofrog - yes, this makes sense to put off to a follow-up bug-fix. I think it might be good to have a little helper function that can also deal with representations.

astrofrog commented on astropy/coordinates/builtin_frames/lsr.py:
```diff
@@ -158,3 +159,93 @@ def galacticlsr_to_galactic(lsr_coord, galactic_frame):
+GORDON1975_V_BARY = 20*u.km/u.s
+GORDON1975_DIRECTION = FK4(ra=270*u.deg, dec=30*u.deg, equinox='B1900')
+
+v_offset_lsrk = ((GORDON1975_V_BARY * GORDON1975_DIRECTION.transform_to(ICRS).data)
```
In fact it would actually be better to simply compute all the offsets in ICRS and hard-code those (not even the Galactic values). Maybe what I'll do is put the final values hard-coded, but leave the code to generate them in comments for reproducibility.

astrofrog replied:
Done

astrofrog commented on astropy/coordinates/builtin_frames/lsr.py:
```diff
@@ -158,3 +159,93 @@ def galacticlsr_to_galactic(lsr_coord, galactic_frame):
+    """
+
+
+V_BARY_DELHAYE1965 = r.CartesianDifferential([9, 12, 7] * u.km/u.s)
```
Done

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+                                 CartesianDifferential,
+                                 CartesianRepresentation, SkyCoord)
+from astropy.coordinates.spectral_quantity import SpectralQuantity
+from astropy.coordinates.baseframe import (BaseCoordinateFrame, FrameMeta,
```
Done

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+from astropy.coordinates.baseframe import (BaseCoordinateFrame, FrameMeta,
+                                           frame_transform_graph)
+from astropy.utils.exceptions import AstropyUserWarning
+from astropy.utils.compat import NUMPY_LT_1_17
```
Done

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+    return (vel1 + vel2) / (1 + vel1 * vel2 / c ** 2)
+
+
+def _apply_relativistic_doppler_shift(self, velocity):
```
Oops - done :)

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+
+    if squantity.unit.is_equivalent(u.m):  # wavelength
+        return squantity * doppler_factor
+    elif squantity.unit.is_equivalent(u.Hz) or squantity.unit.is_equivalent(u.eV) or squantity.unit.is_equivalent(1 / u.m):
```
Fixed

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+    Set the differentials to be stationary on a coordinate object.
+    """
+    coord_diffs = CartesianDifferential(u.Quantity([0, 0, 0] * u.km / u.s))
+    new_data = coord.data.to_cartesian().with_differentials(coord_diffs)
```
No - done!

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+        energy, wavenumber, or velocity values.
+    unit : str or `~astropy.units.Unit`
+        Unit for the given spectral values.
+    observer : `~astropy.coordinates.BaseCoordinateFrame` or `~astropy.coordinates.SkyCoord`, optional
```
Done although see next comment

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,860 @@
+                radial_velocity = getattr(value, 'radial_velocity', None)
+
+        # Validate the observer and target
+        if observer is not None and not isinstance(observer, (SkyCoord, BaseCoordinateFrame)):
```
I'd like to punt on this for now (frames without data as observers) - for the velocity frame transformation it's fine since the position doesn't matter, but for this I think I'd rather we discussed it in a separate PR. In the interest of wrapping this up for feature freeze I'd be tempted to do it as it's implemented right now and then potentially allow these extra use cases in future.

astrofrog commented:
@StuartLittlefair - thanks for the new comments! I addressed them all except two, which I've left unresolved above so you can see my replies. Are you happy to leave those to be resolved after this PR is merged?

StuartLittlefair commented:
@astrofrog yes - no problem at all. they can all be addressed by a separate PR later since we've warned the API will change.

astrofrog commented:
> Likewise, I very much think the .with_ syntax does communicate that the returned object is a new object,

Just to be clear, I didn't imply it didn't return a new object, more that it wasn't clear whether the new object had the same data values, or whether e.g. ``with_redshift`` actually did a redshift *transformation* (it didn't do a transformation) - so that's why I decided to go with replicate for that task, for consistency with other astropy objects.

astrofrog commented:
@mhvk - just to check, are you now happy with this as-is, or are you planning on doing another round of review?

(I believe @eteq is currently reviewing this, so this should not be merged yet)

mhvk commented:
Am hoping to have another look... But OK to merge if that doesn't happen before end of day.

mhvk commented:
@astrofrog - I don't think I'll get around to another round, but as others approved (and it looked very good), I'm fine with it being merged when you're done with implementing comments and CI passes, etc.

astrofrog commented on astropy/coordinates/spectral_coordinate.py:
```diff
@@ -0,0 +1,905 @@
+    radial_velocity : `~astropy.units.Quantity`, optional
+        The radial velocity of the target with respect to the observer.
+    redshift : float, optional
+        The relativistic redshift of the target with respect to the observer.
```
Done

astrofrog commented on astropy/coordinates/spectral_quantity.py:
```diff
@@ -0,0 +1,297 @@
+        elif doppler_convention not in DOPPLER_CONVENTIONS:
+            raise ValueError("doppler_convention should be one of {0}".format('/'.join(sorted(DOPPLER_CONVENTIONS))))
+
+        if self.unit.is_equivalent(si.km / si.s) and unit.is_equivalent(si.km / si.s):
```
Done

astrofrog commented:
@eteq - thanks for the review! A few quick notes:

> Should SpectralQuantity be in units instead of coordinates?

It moved there temporarily during this PR, but it was suggested that it should be moved back, as other classes such as ``Distance`` and so on also live there.

> Somewhere in the docs this should say that it's the relativistic redshift to rv convention, not v=c*z . That is a convention, even if it's not the most "physical" one.

I've added a note to the docs in the latest commit.

> Is a "quick start"/at-a-glance section needed for the SpectralCoord docs?

I don't think it would be easy to add one in the ``spectralcoord.rst`` page since I do think we need to explain things step by step, but I'm opening to someone giving it a try!

> SpectrumQuantity would be quite a bit more convenient if it had the spectral flux density equivalency as a semi-baked in option. To not delay anything I'll just try implementing that as a PR to this PR, but can always change it to against master and include it later.

I'll check your PR - thanks!

mhvk commented:
@eteq, @astrofrog - I'm not sure I understand the use for spectral flux density - would this be if combined with a `Quantity` of flux densities? If it is not totally obvious, I'd go for another PR - this one is very long already (which is mostly why I haven't been able to get back to it).

astrofrog commented:
@mhvk - here is @eteq's PR: https://github.com/astrofrog/astropy/pull/100 - it's more a convenience property to get a spectral density equivalency for that spectral quantity.

mhvk commented:
Ah, that property makes sense, though personally I'd first see what is useful in practice - after all, `fd.to(unit, sq.spectral_density())` is not much less typing than `fd.to(unit, u.spectral_density(sq))`.

eteq commented:
> Ah, that property makes sense, though personally I'd first see what is useful in practice - after all, fd.to(unit, sq.spectral_density()) is not much less typing than fd.to(unit, u.spectral_density(sq)).

OK, I'll re-cast that into a PR against master since I don't want it to block this one.  Since it's clearly had lots of great review at this point and everything as passing, I'm merging this now.  Thanks everyone for all the great work on this!

bsipocz commented:
is specutils compatible with this one, or it's not in the plan to switch to using this straight away?

pllim commented:
Is there a need to also document this in https://docs.astropy.org/en/latest/io/misc.html#schemas ?

Cadair commented:
So this broken SunPy's astropy-dev build FYI. I will investigate.

taldcroft commented:
This broke #10154, though one might reasonably view things the other way round.  I had to change a spectral coord test due to the strict definition of equality now: 3082a81b4dc111.  

I do have to wonder if #10154 will break other tests of code that were explicitly or implicitly using `==` as a substitute for `is`.  The first version of #10154 returned `False` instead of raising when a comparison cannot be done, but based on the way things are going in numpy, we decided that comparing objects that can't be compared should raise an exception.

mhvk commented:
Perhaps we should reconsider raising the exception and instead return `False` after all?  That said, I think the test was not really clean either. And there is a case to be made for making clear to the user why a comparison cannot be done. Anyway, obviously numpy struggled with the same...
